packages/typescript-estree/src/parseSettings/getProjectConfigFiles.ts

```diff
@@ -0,0 +1,79 @@
+import debug from 'debug';
+import * as fs from 'fs';
+import * as path from 'path';
+
+import type { ParseSettings } from '.';
+
+const log = debug('typescript-eslint:typescript-estree:getProjectConfigFiles');
+
+interface TSConfigMatch {
+  exists: boolean;
+  timestamp: number;
+}
+
+/**
+ * How many milliseconds we will respect the cache of an fs.existsSync check.
+ */
+const RECHECK_FILE_THRESHOLD_MS = 50;
+
+const tsconfigMatchCache = new Map<string, TSConfigMatch>();
+
+/**
+ * Checks for a file file on disk, respecting a limited-time cache.
+ * This keeps a caches checked paths with existence and `Date.now()` timestamp.
+ * After `RECHECK_FILE_THRESHOLD_MS`, cache entries are ignored.
+ *
+ * @param filePath File path to check for existence on disk.
+ * @returns Whether the file exists on disk.
+ * @remarks
+ * We don't (yet!) have a way to attach file watchers on disk, but still need to
+ * cache file checks for rapid subsequent calls to fs.existsSync. See discussion
+ * in https://github.com/typescript-eslint/typescript-eslint/issues/101.
+ */
+function existsSyncCached(filePath: string): boolean {
+  const cached = tsconfigMatchCache.get(filePath);
+  const now = Date.now();
+
+  if (cached && now - cached.timestamp < RECHECK_FILE_THRESHOLD_MS) {
+    return cached.exists;
+  }
+
+  const exists = fs.existsSync(filePath);
+
+  tsconfigMatchCache.set(filePath, {
+    exists,
+    timestamp: now,
+  });
+
+  return exists;
+}
+
+export function getProjectConfigFiles(
```

Comment:
do we also want to cache the match for a file:tsconfig pair to save re-traversing?
I don't think it's a terrible thing for us to assume that tsconfigs are static - it's actually what we really do right now anyways if you have a statically defined set of paths.

what is the potential issue? we will keep using a project after you've moved/deleted the tsconfig?

Well here's a funny story - we do that right now. We don't ever delete projects from our cache - which means that after you've deleted your tsconfig, we'll still be using the program we created from it.

There is one way for us to handle the deletion - it's by crashing (haha!)
https://github.com/typescript-eslint/typescript-eslint/blob/b2b8ce2bebb1327ff08ce3d98c7521cd1bdab74c/packages/typescript-estree/src/create-program/getWatchProgramsForProjects.ts#L408-L409
As part of our "maybe invalidation" subroutine we call `fs.statSync` unguarded - if the tsconfig has been moved/deleted, then that'll throw and crash our parser.

This "maybe invalidation" path only occurs if we look for a file and don't find it in any of the `parserOptions.projects`. (i.e. algorithm is (1) check each project, (2) if not found - invalidate a project then check it again).

So ultimately the DevX of "if you move a tsconfig restart your IDE" isn't a terrible thing considering it's a rarer case and it's what users have to do right now.
The perf wins would be worth it IMO.

Reply:
also worth noting that based on the above simplification - another caching point that we could make use of is directory-level caching.
if we parse `src/foo/bar/baz.ts` and find `src/tsconfig.json`, then we know that if we parse any of `src/foo/bar/bam.ts`, `src/foo/bam.ts` or `src/bam.ts`, then the closes tsconfig is always going to be `src/tsconfig.json`!!!

So you can cache not only the file:tsconfig pairs, but also the folder:tsconfig pairs.
I.e. i'd expect a pair of caches that look like:
```ts
const fileResolutonCache = new Map([
  ['src/foo/bar/baz.ts', 'src/tsconfig.json'],
  // ...
]);
const folderResolutionCache = new Map([
  ['src/foo/bar/', 'src/tsconfig.json'],
  ['src/foo/', 'src/tsconfig.json'],
  ['src/', 'src/tsconfig.json'],
]);
```

again the wins here would be huge because it means that once we find the tsconfig for a package, we likely only need to do an `O(1)` lookup to find the tsconfig instead of doing an `O(folder depth)` - which should be much faster given that it shaves off `O(folder depth)` file existence checks!

Reply:
Ah this is great! Much better, agreed 😄 thanks!